src/ImmutableVirtualizedList/ImmutableVirtualizedList.js

```diff
@@ -58,14 +75,40 @@ class ImmutableVirtualizedList extends PureComponent {
 
   static defaultProps = {
     ...VirtualizedList.defaultProps,
-
+    treatAsSections: false,
     renderEmptyInList: 'No data.',
   };
 
+  state = {
+    flattenedData: this.props.treatAsSections
+      ? utils.flattenMap(this.props.immutableData)
+      : null,
+  }
+
+  componentDidUpdate(prevProps) {
+    const { treatAsSections, immutableData } = this.props;
+    if (treatAsSections && prevProps.immutableData !== immutableData) {
+      const flattenedData = utils.flattenMap(immutableData);
+      this.setState({
+        flattenedData,
```

Comment:
@cooperka I chose to hold `flattenedData` in `state` so that it is only created once on each update to `immutableData`, as the flattening transformation will be fairly expensive.

Reply:
One alternative, to avoid using `state` at all, would be to use a `WeakMap` to cache the results of the flattening transformation. As we are working with immutable data this would be easy to reason about.

Reply:
I think using state is fine, but if you put it in `willUpdate` instead of `didUpdate` I think it will be more performant (that's why the linter [flagged this](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-did-update-set-state.md)).

Reply:
I chose `didUpdate` because in the [react docs for `componentWillUpdate()`](https://reactjs.org/docs/react-component.html#componentwillupdate) it says "Note that you cannot call `this.setState()` here; nor should you do anything else (e.g. dispatch a Redux action) that would trigger an update to a React component before `componentWillUpdate()` returns."

Reply:
Sorry, `componentWillReceiveProps` is what you should use here! Not `willUpdate`.